File: app/exporter/settlement_report_exporter.py
"""月度结算表导出器：仿模板生成一个多 sheet Excel（每人一个 sheet + 公共费用）

结构（仿「2025年1月结算表.xlsx」）：
- 列：序号 | 项目 | 本期 | 本年累计 | 备注
- 行：一 上年结余结转 / 二 本月收款金额（1.本月开票本月收款 2.收回以前应收款）
     / 三 本月开具发票金额（1.本月开票已收款 2.本月开票未收款）
     / 四 未收款金额（本期=本月未收；累计=本年累计未收）
     / 五 业务收入 / 六 减：分成报酬及费用（逐类型）
     / 七 减：税金及会费（增值税、城建及附加=开票净额÷1.06×6%×1.12；抵扣进项留空）
     / 结余 = 五 - 六 - 七
备注：收回以前应收款>0 时自动生成
"""
from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import Dict

import openpyxl
from openpyxl.styles import Alignment, Border, Font, Side
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.page import PageMargins
from openpyxl.worksheet.properties import PageSetupProperties

from app.engine.person_settlement import build_settlement

THIN = Side(style="thin", color="999999")
BORDER = Border(left=THIN, right=THIN, top=THIN, bottom=THIN)
BOLD = Font(bold=True)
CENTER = Alignment(horizontal="center", vertical="center")

CN_NUM = ["一", "二", "三", "四", "五", "六", "七"]


def _tax(amount: float) -> float:
    """增值税、城建及附加 = 开票净额 ÷ 1.06 × 6% × 1.12"""
    return round(amount / 1.06 * 0.06 * 1.12, 2)


def _chinese_title(month: int) -> str:
    cn = "一二三四五六七八九十"
    if month <= 10:
        return cn[month - 1] + "月"
    return "十" + (cn[month - 11] if month > 11 else "") + "月"


def _month_total(st: Dict, month: int, keys) -> float:
    return round(sum(st["months"][month][k] for k in keys), 2)


def report_note_rec(st: Dict, month: int) -> str:
    """收款类备注（写在「二、本月收款金额」行）"""
    m = st["months"]
    notes = []
    if m[month]["rec_cur_year"] > 0.01:
        notes.append(f"本月收回本年应收款{m[month]['rec_cur_year']:,.2f}元")
    if m[month]["rec_prev_year"] > 0.01:
        notes.append(f"本月收回上年应收款{m[month]['rec_prev_year']:,.2f}元")
    if m[month]["rec_refund_cur"] < -0.01:
        notes.append(f"本月退本年应收款{abs(m[month]['rec_refund_cur']):,.2f}元")
    if m[month]["rec_refund_prev"] < -0.01:
        notes.append(f"本月退上年应收款{abs(m[month]['rec_refund_prev']):,.2f}元")
    cum_prev_year = sum(m[mo]["rec_prev_year"] for mo in range(1, month + 1))
    cum_refund_prev = sum(m[mo]["rec_refund_prev"] for mo in range(1, month + 1))
    if cum_prev_year > 0.01:
        notes.append(f"本年累计收回上年应收款{cum_prev_year:,.2f}元")
    if cum_refund_prev < -0.01:
        notes.append(f"本年累计退上年应收款{abs(cum_refund_prev):,.2f}元")
    return "\n".join(notes)


def report_note_inv(st: Dict, month: int) -> str:
    """开票类备注（写在「三、本月开具发票金额」行）"""
    m = st["months"]
    notes = []
    if m[month]["inv_red_cur"] < -0.01:
        notes.append(f"本月红冲本年{abs(m[month]['inv_red_cur']):,.2f}元")
    if m[month]["inv_red_prev"] < -0.01:
        notes.append(f"本月红冲上年{abs(m[month]['inv_red_prev']):,.2f}元")
    cum_red_prev = sum(m[mo]["inv_red_prev"] for mo in range(1, month + 1))
    if cum_red_prev < -0.01:
        notes.append(f"本年累计红冲上年{abs(cum_red_prev):,.2f}元")
    return "\n".join(notes)


def build_report_rows(st: Dict, year: int, month: int) -> list:
    """结算表行数据：[(seq, name, cur, total, bold), ...]（预览与导出共用）"""
    m = st["months"]
    months = list(range(1, 13))
    cum = lambda vals12: round(sum(vals12[:month]), 2)  # noqa: E731
    rows = []

    def row(seq, name, cur, total, bold=False):
        rows.append([seq, name, None if cur is None else round(cur, 2),
                     None if total is None else round(total, 2), bold])

    # 一、上年结余结转
    row("一", "上年结余结转", None, None)
    # 二、本月收款金额
    rec_keys = ["rec_open_cur", "rec_cur_year", "rec_prev_year", "rec_refund_cur", "rec_refund_prev"]
    cur2 = _month_total(st, month, rec_keys)
    tot2 = round(sum(_month_total(st, mo, rec_keys) for mo in months[:month]), 2)
    row("二", "本月收款金额", cur2, tot2, bold=True)
    row("1", "本月开票本月收款", m[month]["rec_open_cur"],
        cum([m[mo]["rec_open_cur"] for mo in months]))
    row("2", "收回以前应收款", round(m[month]["rec_cur_year"] + m[month]["rec_prev_year"], 2),
        cum([m[mo]["rec_cur_year"] + m[mo]["rec_prev_year"] for mo in months]))
    # 三、本月开具发票金额
    cur3 = m[month]["inv_total"]
    tot3 = round(sum(m[mo]["inv_total"] for mo in months[:month]), 2)
    row("三", "本月开具发票金额", cur3, tot3, bold=True)
    row("1", "本月开票已收款", m[month]["inv_open_received"],
        cum([m[mo]["inv_open_received"] for mo in months]))
    row("2", "本月开票未收款", m[month]["inv_open_uncollected"],
        cum([m[mo]["inv_open_uncollected"] for mo in months]))
    # 四、未收款金额
    row("四", "未收款金额", st["uncollected_month"][month], st["uncollected_total"], bold=True)
    # 五、业务收入
    row("五", "业务收入", m[month]["income"], cum([m[mo]["income"] for mo in months]), bold=True)
    # 六、减：分成报酬及费用（排除折旧/银行结息/城建税及附加——模板口径）
    _EXCLUDE = ("折旧", "银行结息", "城建税", "教育附加")
    exp = {t: v for t, v in st["expenses"].items() if not any(k in t for k in _EXCLUDE)}
    cur6 = round(sum(exp.get(t, {}).get(month, 0.0) for t in exp), 2)
    tot6 = round(sum(exp.get(t, {}).get(mo, 0.0) for t in exp for mo in months[:month]), 2)
    row("六", "减：分成报酬及费用", cur6, tot6, bold=True)
    _EXP_ORDER = ["分成报酬", "合办人员报酬", "社保", "刷卡汽油费", "发票报销", "停车费",
                  "高温费", "旅游费", "行政工资", "行政年终奖", "实习工资", "公积金"]
    exp_types = sorted(exp.keys(), key=lambda t: (_EXP_ORDER.index(t) if t in _EXP_ORDER else 99, t))
    for i, etype in enumerate(exp_types, 1):
        row(str(i), etype, exp[etype].get(month, 0.0), cum([exp[etype].get(mo, 0.0) for mo in months]))
    # 七、减：税金及会费
    tax_cur = _tax(max(cur3, 0.0))
    tax_tot = _tax(max(tot3, 0.0))
    row("七", "减：税金及会费", round(tax_cur, 2), round(tax_tot, 2), bold=True)
    row("1", "增值税、城建及附加", tax_cur, tax_tot)
    row("2", "抵扣进项", None, None)
    # 结余
    row("", "结余", round(cur5 := m[month]["income"] - cur6 - tax_cur, 2),
        round(tot5 := cum([m[mo]["income"] for mo in months]) - tot6 - tax_tot, 2), bold=True)
    return rows


def _write_sheet(ws, person: str, st: Dict, year: int, month: int) -> None:
    m = st["months"]
    months = list(range(1, 13))

    # 标题
    ws.merge_cells("A1:E1")
    ws["A1"] = "浙江震天律师事务所"
    ws["A1"].font = Font(size=13, bold=True)
    ws["A1"].alignment = CENTER
    ws["B2"] = f"姓名：{person}"
    ws["C2"] = f"二〇{'{:02d}'.format(year % 100)}年{_chinese_title(month)}结算表"

    # 表头
    for j, h in enumerate(["序号", "项目", "本期", "本年累计", "备注"], 1):
        c = ws.cell(3, j, h)
        c.font = BOLD
        c.alignment = CENTER
        c.border = BORDER

    for seq, name, cur, total, bold in build_report_rows(st, year, month):
        r = ws.max_row + 1
        ws.cell(r, 1, seq).alignment = CENTER
        ws.cell(r, 2, name).alignment = Alignment(horizontal="left", vertical="center")
        c3 = ws.cell(r, 3, cur)
        c4 = ws.cell(r, 4, total)
        for c in (c3, c4):
            if c.value is not None:
                c.number_format = "#,##0.00"
                c.alignment = Alignment(horizontal="right", vertical="center")
        for j in range(1, 6):
            ws.cell(r, j).border = BORDER
        if bold:
            ws.cell(r, 2).font = BOLD
            for j in (3, 4):
                ws.cell(r, j).font = BOLD
        # 备注按类别写在对应行（二收款 / 三开票）
        if name == "本月收款金额":
            note = report_note_rec(st, month)
            if note:
                c = ws.cell(r, 5, note)
                c.alignment = Alignment(wrap_text=True, vertical="center")
                ws.row_dimensions[r].height = max(ws.row_dimensions[r].height or 0, 50)
        elif name == "本月开具发票金额":
            note = report_note_inv(st, month)
            if note:
                c = ws.cell(r, 5, note)
                c.alignment = Alignment(wrap_text=True, vertical="center")
                ws.row_dimensions[r].height = max(ws.row_dimensions[r].height or 0, 50)

    # ---- 自动列宽（按内容，汉字计 2 字符）----
    def _char_w(text: str) -> int:
        return sum(2 if ord(ch) > 127 else 1 for ch in str(text or ""))

    # A 列序号固定窄宽；E 列备注保底 28（无备注时也美观）
    ws.column_dimensions["A"].width = 6
    ws.column_dimensions["E"].width = 28
    for col_idx in (2, 3, 4):
        maxw = 0
        for row in ws.iter_rows(min_col=col_idx, max_col=col_idx):
            for cell in row:
                if cell.value is None:
                    continue
                lines = str(cell.value).split("\n")
                maxw = max(maxw, max(_char_w(line) for line in lines))
        ws.column_dimensions[get_column_letter(col_idx)].width = min(maxw + 2, 48)
    # E 列备注若内容更长则扩展
    maxw_e = 0
    for row in ws.iter_rows(min_col=5, max_col=5):
        for cell in row:
            if cell.value is None:
                continue
            maxw_e = max(maxw_e, max(_char_w(line) for line in str(cell.value).split("\n")))
    ws.column_dimensions["E"].width = min(max(28, maxw_e + 2), 60)

    # ---- 打印设置：每个 sheet 打印在一页 ----
    ws.page_setup.orientation = "portrait"
    ws.page_setup.fitToWidth = 1
    ws.page_setup.fitToHeight = 1
    ws.sheet_properties.pageSetUpPr = PageSetupProperties(fitToPage=True)
    ws.page_margins = PageMargins(left=0.25, right=0.25, top=0.3, bottom=0.3)


def export_report(out_path: str | Path, year: int, month: int, persons: list[str]) -> Path:
    """生成月度结算表（多 sheet），persons 为要导出的员工名单（含"公共费用"）"""
    data = build_settlement(year)
    wb = openpyxl.Workbook()
    wb.remove(wb.active)
    for person in persons:
        # 模板 sheet 名为"公共费用"，引擎中费用经办人为"公共"
        st = data.get(person)
        if st is None and person == "公共费用":
            st = data.get("公共")
        if st is None:
            st = {
                "staff_type": "其他",
                "months": {mo: {k: 0.0 for k in (
                    "rec_open_cur", "rec_cur_year", "rec_prev_year", "rec_refund_cur",
                    "rec_refund_prev", "inv_open_received", "inv_open_uncollected",
                    "inv_red_cur", "inv_red_prev", "inv_total", "income")} for mo in range(1, 13)},
                "uncollected_month": {mo: 0.0 for mo in range(1, 13)},
                "uncollected_total": 0.0,
                "expenses": {},
            }
        ws = wb.create_sheet(title=person)
        _write_sheet(ws, person, st, year, month)
    out = Path(out_path)
    out.parent.mkdir(parents=True, exist_ok=True)
    wb.save(out)
    return out

File: csharp/LawFirm.Exporter/SettlementReportExporter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using Microsoft.Data.Sqlite;
using NPOI.SS.UserModel;
using NPOI.SS.Util;
using NPOI.XSSF.UserModel;

namespace LawFirm.Exporter;

/// <summary>
/// 月度结算表导出器（Pilot T2，C# / NPOI 端口）。
///
/// 逐格镜像 Python app/exporter/settlement_report_exporter.py 的 _write_sheet /
/// build_report_rows / report_note_*，写入 xlsx 供 diff_xlsx.py 逐格比对。
///
/// 对齐重点（对照 diff_xlsx.py 的 7 个维度）：
///  - sheet 名/顺序：每人 合伙/聘用/兼职(有数据才出) + 汇总 + 公共费用，与 Python 完全一致；
///  - 单元格值 + number_format：仅数值格设 "#,##0.00"，空值保持 General；
///  - 合并单元格：仅 A1:E1（标题）；
///  - 表头序列：第 3 行 = 序号|项目|本期|本年累计|备注；
///  - 汇总行：含「结余」关键词的行逐格对齐；
///  - 页面设置：orientation=portrait、fitToPage=1、四边 margins=0.25/0.25/0.3/0.3、
///    header/footer=0.5/0.5（fitToWidth/fitToHeight 因 NPOI DefaultValue(1) 不会落盘，
///    语义等价，详见 README_T2 第 5 节）。
/// （列宽/行高 diff 不比较，故不强制对齐，仅做基本可用性设置。）
/// </summary>
public static class SettlementReportExporter
{
    private static readonly int[] AllMonths = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 };
    private static readonly string[] Header = { "序号", "项目", "本期", "本年累计", "备注" };

    public static string ExportReport(
        SqliteConnection conn, string outPath, int year, int month, List<string> persons)
    {
        // 输出目录常常不存在（csharp/out 已被 .gitignore 忽略，克隆/同步后不会自带），
        // 不先建目录 FileStream 会抛 DirectoryNotFoundException（"找不到路径…的一部分"）。
        var outDir = Path.GetDirectoryName(Path.GetFullPath(outPath));
        if (!string.IsNullOrEmpty(outDir)) Directory.CreateDirectory(outDir);

        if (persons == null || persons.Count == 0)
        {
            // 与 Python 导出器一致：空名单兜底为合法占位 xlsx（>=1 sheet），避免无效 workbook。
            // 注意：此处变量名不能与外层作用域的 wb / fs 同名 —— C# 规则 CS0136 禁止在嵌套
            // 作用域声明与封闭作用域同名的局部变量（与声明先后顺序无关），故加 Empty 后缀。
            var wbEmpty = new XSSFWorkbook();
            var wsEmpty = wbEmpty.CreateSheet("无数据");
            wsEmpty.CreateRow(0).CreateCell(0).SetCellValue("无结算数据");
            wsEmpty.CreateRow(1).CreateCell(0).SetCellValue($"{year}年无结算人员数据，未生成结算表。");
            // 与 Python golden 对齐页边距：Python 的占位 sheet 没有走 _write_sheet，
            // 因此沿用 openpyxl 的默认页边距（left/right=0.75、top/bottom=1.0）；
            // 而 NPOI 的默认是 0.7/0.7/0.75/0.75，若不显式写出，逐格 diff 会在
            // page_setup 维度报 4 处不一致。此处显式写成 Python 侧的实际值。
            // （方向/缩放两边默认一致，已验证无差异，故不动以免引入新的不一致。）
            wsEmpty.SetMargin(MarginType.LeftMargin, 0.75f);
            wsEmpty.SetMargin(MarginType.RightMargin, 0.75f);
            wsEmpty.SetMargin(MarginType.TopMargin, 1.0f);
            wsEmpty.SetMargin(MarginType.BottomMargin, 1.0f);
            Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(outPath))!);
            using var fsEmpty = new FileStream(outPath, FileMode.Create, FileAccess.Write);
            wbEmpty.Write(fsEmpty);
            return outPath;
        }
        var dataAll = SettlementEngine.Build(conn, year);
        var ordered = ExpenseCatHelper.OrderedTypes(conn);
        var wb = new XSSFWorkbook();
        foreach (var person in persons)
        {
            if (person == "公共费用")
            {
                var st = dataAll.TryGetValue("公共", out var g) ? g : new PersonSettlement();
                var ws = wb.CreateSheet("公共费用");
                WriteSheet(ws, "公共费用", st, year, month, ordered);
                continue;
            }
            bool anyData = false;
            foreach (var ptype in new[] { "合伙", "聘用", "兼职" })
            {
                var st = SettlementEngine.Build(conn, year, person, ptype).GetValueOrDefault(person);
                if (st != null && HasData(st, month))
                {
                    var ws = wb.CreateSheet(person + ptype);
                    WriteSheet(ws, person, st, year, month, ordered);
                    anyData = true;
                }
            }
            var stAll = dataAll.GetValueOrDefault(person);
            if (stAll != null && HasData(stAll, month))
            {
                var ws = wb.CreateSheet(person + "汇总");
                WriteSheet(ws, person, stAll, year, month, ordered);
                anyData = true;
            }
            if (!anyData)
            {
                var ws = wb.CreateSheet(person + "汇总");
                WriteSheet(ws, person, new PersonSettlement(), year, month, ordered);
            }
        }
        Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(outPath))!);
        using var fs = new FileStream(outPath, FileMode.Create, FileAccess.Write);
        wb.Write(fs);
        return outPath;
    }

    // ------------------------------------------------------------------
    private static bool HasData(PersonSettlement st, int month)
    {
        var m = st.Months[month];
        double[] vals =
        {
            m.RecOpenCur, m.RecCurYear, m.RecPrevYear, m.RecRefundCur, m.RecRefundPrev,
            m.InvOpenReceived, m.InvOpenUncollected, m.InvRedCur, m.InvRedPrev, m.Income,
        };
        if (vals.Any(v => Math.Abs(v) > 0.01)) return true;
        if (st.UncollectedMonth.GetValueOrDefault(month, 0) > 0.01) return true;
        if (st.Expenses.Values.Sum(e => e.GetValueOrDefault(month, 0.0)) > 0.01) return true;
        return false;
    }

    private static void WriteSheet(
        ISheet ws, string person, PersonSettlement st, int year, int month, List<string> ordered)
    {
        var wbk = ws.Workbook;
        var thin = BorderStyle.Thin;

        ICellStyle Make(bool bold, HorizontalAlignment align, bool numeric)
        {
            var s = wbk.CreateCellStyle();
            s.BorderTop = thin; s.BorderBottom = thin; s.BorderLeft = thin; s.BorderRight = thin;
            s.Alignment = align;
            if (numeric) s.DataFormat = wbk.CreateDataFormat().GetFormat("#,##0.00");
            if (bold) { var f = wbk.CreateFont(); f.IsBold = true; s.SetFont(f); }
            return s;
        }
        var centerText = Make(false, HorizontalAlignment.Center, false);
        var centerTextBold = Make(true, HorizontalAlignment.Center, false);
        var leftText = Make(false, HorizontalAlignment.Left, false);
        var leftTextBold = Make(true, HorizontalAlignment.Left, false);
        var num = Make(false, HorizontalAlignment.Right, true);
        var numBold = Make(true, HorizontalAlignment.Right, true);

        // 标题（A1，合并 A1:E1）
        var r0 = ws.CreateRow(0);
        var cA1 = r0.CreateCell(0);
        cA1.SetCellValue("浙江震天律师事务所");
        var titleFont = wbk.CreateFont();
        titleFont.IsBold = true;
        titleFont.FontHeightInPoints = 13;
        var titleStyle = wbk.CreateCellStyle();
        titleStyle.SetFont(titleFont);
        titleStyle.Alignment = HorizontalAlignment.Center;
        cA1.CellStyle = titleStyle;
        ws.AddMergedRegion(new CellRangeAddress(0, 0, 0, 4));

        // 第二行：姓名 / 期间
        var r1 = ws.CreateRow(1);
        r1.CreateCell(1).SetCellValue($"姓名：{person}");
        r1.CreateCell(2).SetCellValue($"二〇{year % 100:00}年{ChineseTitle(month)}结算表");

        // 表头（第 3 行）
        var r2 = ws.CreateRow(2);
        for (int j = 0; j < 5; j++)
        {
            var c = r2.CreateCell(j);
            c.SetCellValue(Header[j]);
            c.CellStyle = centerTextBold;
        }

        // 数据行
        foreach (var (seq, name, cur, total, bold) in BuildReportRows(st, year, month, ordered))
        {
            int rr = ws.LastRowNum + 1;
            var row = ws.CreateRow(rr);
            var ca = row.CreateCell(0);
            // seq 可能为 ""（结余行）。openpyxl 写空串时落盘为「无值单元格」，
            // 而 NPOI 的 SetCellValue("") 会落盘成空字符串，diff 会读成 '' ≠ None。
            // 因此空 seq 只建单元格套样式、不写值，保持与 golden 一致。
            if (!string.IsNullOrEmpty(seq)) ca.SetCellValue(seq);
            ca.CellStyle = bold ? centerTextBold : centerText;
            var cb = row.CreateCell(1);
            cb.SetCellValue(name);
            cb.CellStyle = bold ? leftTextBold : leftText;
            if (cur.HasValue)
            {
                var cc = row.CreateCell(2);
                cc.SetCellValue(Math.Round(cur.Value, 2));
                cc.CellStyle = bold ? numBold : num;
            }
            if (total.HasValue)
            {
                var cd = row.CreateCell(3);
                cd.SetCellValue(Math.Round(total.Value, 2));
                cd.CellStyle = bold ? numBold : num;
            }
            if (name == "本月收款金额")
            {
                var note = ReportNoteRec(st, month);
                if (!string.IsNullOrEmpty(note)) row.CreateCell(4).SetCellValue(note);
            }
            else if (name == "本月开具发票金额")
            {
                var note = ReportNoteInv(st, month);
                if (!string.IsNullOrEmpty(note)) row.CreateCell(4).SetCellValue(note);
            }
        }

        // 页面设置（对照 diff_xlsx 第 6 维）
        ws.PrintSetup.Landscape = false;     // portrait
        ws.FitToPage = true;
        // NPOI 的 IPrintSetup 属性名为 FitWidth / FitHeight（对应 POI 的 setFitWidth/setFitHeight，
        // 类型为 short），不存在 FitToWidth / FitToHeight —— 后者是 openpyxl 的属性名。
        ws.PrintSetup.FitWidth = (short)1;
        ws.PrintSetup.FitHeight = (short)1;
        // NPOI 2.7.2 的 CT_PageSetup.fitToWidth/fitToHeight 带 [DefaultValue(1)]，
        // XmlSerializer 会在「等于默认值」时省略该属性 —— 与 openpyxl 显式写 1 语义等价，
        // diff_xlsx 侧按 schema 默认值归一（None 视为 1），此处无需再处理。
        // 页边距必须用 double 字面量：写 0.3f（float）会存成 0.30000001192092896。
        ws.SetMargin(MarginType.LeftMargin, 0.25);
        ws.SetMargin(MarginType.RightMargin, 0.25);
        ws.SetMargin(MarginType.TopMargin, 0.3);
        ws.SetMargin(MarginType.BottomMargin, 0.3);
        ws.SetMargin(MarginType.HeaderMargin, 0.5);
        ws.SetMargin(MarginType.FooterMargin, 0.5);
    }

    // ------------------------------------------------------------------
    // 行数据（镜像 build_report_rows）
    // ------------------------------------------------------------------
    private static List<(string seq, string name, double? cur, double? total, bool bold)> BuildReportRows(
        PersonSettlement st, int year, int month, List<string> ordered)
    {
        var m = st.Months[month];
        double Cum(Func<int, double> f) => Math.Round(AllMonths.Take(month).Sum(f), 2);
        var rows = new List<(string, string, double?, double?, bool)>();
        void Row(string seq, string name, double? cur, double? total, bool bold = false)
            => rows.Add((seq, name,
                cur is double c ? Math.Round(c, 2) : (double?)null,
                total is double t ? Math.Round(t, 2) : (double?)null,
                bold));

        Row("一", "上年结余结转", null, null);
        var recKeys = new Func<int, double>[]
        {
            mm => st.Months[mm].RecOpenCur, mm => st.Months[mm].RecCurYear, mm => st.Months[mm].RecPrevYear,
            mm => st.Months[mm].RecRefundCur, mm => st.Months[mm].RecRefundPrev,
        };
        double MonthTotal(Func<int, double>[] ks) => Math.Round(ks.Sum(k => k(month)), 2);
        double Cur2 = MonthTotal(recKeys);
        double Tot2 = Math.Round(AllMonths.Take(month).Sum(mm => recKeys.Sum(k => k(mm))), 2);
        Row("二", "本月收款金额", Cur2, Tot2, true);
        Row("1", "本月开票本月收款", m.RecOpenCur, Cum(mm => st.Months[mm].RecOpenCur));
        Row("2", "收回以前应收款", Math.Round(m.RecCurYear + m.RecPrevYear, 2),
            Cum(mm => st.Months[mm].RecCurYear + st.Months[mm].RecPrevYear));
        double Cur3 = m.InvTotal;
        double Tot3 = Math.Round(AllMonths.Take(month).Sum(mm => st.Months[mm].InvTotal), 2);
        Row("三", "本月开具发票金额", Cur3, Tot3, true);
        Row("1", "本月开票已收款", m.InvOpenReceived, Cum(mm => st.Months[mm].InvOpenReceived));
        Row("2", "本月开票未收款", m.InvOpenUncollected, Cum(mm => st.Months[mm].InvOpenUncollected));
        Row("四", "未收款金额", st.UncollectedMonth.GetValueOrDefault(month, 0), st.UncollectedTotal, true);
        Row("五", "业务收入", m.Income, Cum(mm => st.Months[mm].Income), true);

        // 六、减：分成报酬及费用（排除折旧/银行结息/城建税及附加）
        string[] exclude = { "折旧", "银行结息", "城建税", "教育附加" };
        var exp = st.Expenses
            .Where(kv => !exclude.Any(ex => kv.Key.Contains(ex)))
            .ToDictionary(kv => kv.Key, kv => kv.Value);
        double Cur6 = Math.Round(exp.Values.Sum(e => e.GetValueOrDefault(month, 0)), 2);
        double Tot6 = Math.Round(exp.Values.Sum(e => AllMonths.Take(month).Sum(mm => e.GetValueOrDefault(mm, 0))), 2);
        Row("六", "减：分成报酬及费用", Cur6, Tot6, true);

        var expOrder = ordered.Select((t, i) => (t, i)).ToDictionary(x => x.t, x => x.i);
        var expTypes = exp.Keys
            .OrderBy(t => expOrder.GetValueOrDefault(t, 999))
            .ThenBy(t => t, StringComparer.Ordinal)
            .ToList();
        int idx = 1;
        foreach (var etype in expTypes)
        {
            Row(idx.ToString(), etype,
                exp[etype].GetValueOrDefault(month, 0),
                Cum(mm => exp[etype].GetValueOrDefault(mm, 0)));
            idx++;
        }

        double taxCur = Tax(Math.Max(Cur3, 0));
        double taxTot = Tax(Math.Max(Tot3, 0));
        Row("七", "减：税金及会费", Math.Round(taxCur, 2), Math.Round(taxTot, 2), true);
        Row("1", "增值税、城建及附加", taxCur, taxTot);
        Row("2", "抵扣进项", null, null);

        double cur5 = m.Income - Cur6 - taxCur;
        double tot5 = Cum(mm => st.Months[mm].Income) - Tot6 - taxTot;
        Row("", "结余", Math.Round(cur5, 2), Math.Round(tot5, 2), true);
        return rows;
    }

    private static double Tax(double amount) => Math.Round(amount / 1.06 * 0.06 * 1.12, 2);

    private static string ChineseTitle(int month)
    {
        const string cn = "一二三四五六七八九十";
        if (month <= 10) return cn[month - 1] + "月";
        return "十" + (month > 11 ? cn[month - 11].ToString() : "") + "月";
    }

    private static string Fmt(double x) => x.ToString("#,##0.00", CultureInfo.InvariantCulture);

    private static string ReportNoteRec(PersonSettlement st, int month)
    {
        var m = st.Months[month];
        var notes = new List<string>();
        if (m.RecCurYear > 0.01) notes.Add($"本月收回本年应收款{Fmt(m.RecCurYear)}元");
        if (m.RecPrevYear > 0.01) notes.Add($"本月收回上年应收款{Fmt(m.RecPrevYear)}元");
        if (m.RecRefundCur < -0.01) notes.Add($"本月退本年应收款{Fmt(Math.Abs(m.RecRefundCur))}元");
        if (m.RecRefundPrev < -0.01) notes.Add($"本月退上年应收款{Fmt(Math.Abs(m.RecRefundPrev))}元");
        double cumPrevYear = AllMonths.Take(month).Sum(mm => st.Months[mm].RecPrevYear);
        double cumRefundPrev = AllMonths.Take(month).Sum(mm => st.Months[mm].RecRefundPrev);
        if (cumPrevYear > 0.01) notes.Add($"本年累计收回上年应收款{Fmt(cumPrevYear)}元");
        if (cumRefundPrev < -0.01) notes.Add($"本年累计退上年应收款{Fmt(Math.Abs(cumRefundPrev))}元");
        return string.Join("\n", notes);
    }

    private static string ReportNoteInv(PersonSettlement st, int month)
    {
        var m = st.Months[month];
        var notes = new List<string>();
        if (m.InvRedCur < -0.01) notes.Add($"本月红冲本年{Fmt(Math.Abs(m.InvRedCur))}元");
        if (m.InvRedPrev < -0.01) notes.Add($"本月红冲上年{Fmt(Math.Abs(m.InvRedPrev))}元");
        double cumRedPrev = AllMonths.Take(month).Sum(mm => st.Months[mm].InvRedPrev);
        if (cumRedPrev < -0.01) notes.Add($"本年累计红冲上年{Fmt(Math.Abs(cumRedPrev))}元");
        return string.Join("\n", notes);
    }
}
